clients/da-vinci-client/src/main/java/com/linkedin/davinci/notifier/DaVinciPushStatusUpdateTask.java

```diff
@@ -57,75 +63,155 @@ public void batchPushEndSignalSent() {
+    handleBatchPushStatus();
+    handleIncrementalPushStatus();
+  }
+
+  private void handleBatchPushStatus() {
     if (!isBatchPushStartSignalSent()) {
-      pushStatusStoreWriter.writeVersionLevelPushStatus(
-          version.getStoreName(),
-          version.getNumber(),
-          ExecutionStatus.STARTED,
-          getTrackedPartitions());
+      sendPushStatus(ExecutionStatus.STARTED, Optional.empty());
       batchPushStartSignalSent();
-    } else if (!isBatchPushEndSignalSent() && areAllPartitionsOnSameTerminalStatus(ExecutionStatus.COMPLETED)
+    } else if (!isBatchPushEndSignalSent()
+        && areAllPartitionsOnSameTerminalStatus(ExecutionStatus.COMPLETED, Optional.empty())
         && areAllPartitionFuturesCompletedSuccessfully.get()) {
-      pushStatusStoreWriter.writeVersionLevelPushStatus(
-          version.getStoreName(),
-          version.getNumber(),
-          ExecutionStatus.COMPLETED,
-          getTrackedPartitions());
+      sendPushStatus(ExecutionStatus.COMPLETED, Optional.empty());
       batchPushEndSignalSent();
-      // Shutdown the scheduler after sending the final status
-      shutdown();
-    } else if (!isBatchPushEndSignalSent() && isAnyPartitionOnErrorStatus()) {
-      pushStatusStoreWriter.writeVersionLevelPushStatus(
-          version.getStoreName(),
-          version.getNumber(),
-          ExecutionStatus.ERROR,
-          getTrackedPartitions());
+      batchPushPartitionStatus.clear();
+    } else if (!isBatchPushEndSignalSent() && isAnyPartitionOnErrorStatus(Optional.empty())) {
+      sendPushStatus(ExecutionStatus.ERROR, Optional.empty());
       batchPushEndSignalSent();
-      // Shutdown the scheduler after sending the final status
-      shutdown();
+      batchPushPartitionStatus.clear();
     }
   }
 
-  /**
-   * Get the partition id set that is being tracked
-   */
-  public Set<Integer> getTrackedPartitions() {
-    return partitionStatus.keySet();
+  private void handleIncrementalPushStatus() {
+    incrementalPushVersionToStatus.forEach((incrementalPushVersion, incrementalPushStatus) -> {
+      if (!incrementalPushStatus.startSignalSent) {
+        sendPushStatus(ExecutionStatus.START_OF_INCREMENTAL_PUSH_RECEIVED, Optional.of(incrementalPushVersion));
+        incrementalPushStatus.startSignalSent = true;
+      } else if (!incrementalPushStatus.endSignalSent && areAllPartitionsOnSameTerminalStatus(
+          ExecutionStatus.END_OF_INCREMENTAL_PUSH_RECEIVED,
+          Optional.of(incrementalPushVersion))) {
+        sendPushStatus(ExecutionStatus.END_OF_INCREMENTAL_PUSH_RECEIVED, Optional.of(incrementalPushVersion));
+        incrementalPushStatus.endSignalSent = true;
+        // Clean up the status map for incremental push versions in case there are too many of them and cause OOM
+        incrementalPushStatus.partitionStatus.clear();
+      } else if (!incrementalPushStatus.endSignalSent
+          && isAnyPartitionOnErrorStatus(Optional.of(incrementalPushVersion))) {
```

Comment:
For this branch I am a bit confused. Will we report ERROR on specific incremental push version?

Reply:
I believe we do, as long as COMPLETED is not reported yet.
You can check all the invocations of
`StoreIngestionTask#handleIngestionException(Exception e)`
`StoreIngestionTask#handleIngestionThrowable(Throwable e)`
`LeaderFollowerStoreIngestionTask#waitForLastLeaderPersistFuture`

I should probably add a check to see if COMPLETED is reported.

Reply:
Double checked, ERROR is never reported with incremental push keys. Before COMPLETED is reported, if ERROR happens, the batch version would be marked as ERROR already.